File: Client/Trivia/Source/Codec/S2C/Objects/PlayerResult.cs
using System;
using Trivia.Models.Raw;

namespace Trivia.Codec.S2C.Objects;

public record PlayerResult
{
    public required User User { get; init; }
    public required int CorrectAnswerCount { get; init; }
    public required int AverageAnswerTimeSecs { get; init; }
    public required int PlaytimeSecs { get; init; }
    public required int Points { get; init; }
    
    public TimeSpan AverageAnswerTime => TimeSpan.FromSeconds(AverageAnswerTimeSecs);
    public TimeSpan Playtime => TimeSpan.FromSeconds(PlaytimeSecs);
};

File: Server/src/common/infrastructure/PlayerResult.cpp
#include "PlayerResult.h"

PlayerResult::PlayerResult(
    const LoggedUser &user,
    const int playtimeSecs,
    const int correctAnswerCount,
    const int averageAnswerTimeSecs,
    const int points
) :
    user(user),
    playtimeSecs(playtimeSecs),
    correctAnswerCount(correctAnswerCount),
    averageAnswerTimeSecs(averageAnswerTimeSecs),
    points(points)
{}

PlayerResult::PlayerResult(const LoggedUser& user, const GameData &data) :
    PlayerResult(
        user,
        data.getPlaytime().count(),
        data.correctAnswerCount,
        data.getAverageAnswerTime().count(),
        data.points
    )
{}
